src/js/service/VendingMachine.js

```diff
@@ -0,0 +1,141 @@
+import { ALERT_MESSAGE, VENDING_MACHINE_CONSTANT } from './constant.js';
+import { removeSpace } from '../util/string.js';
+import { isAmountValid, isInsertedCoinsValid, isNameValid, isPriceValid } from './validator.js';
+import ValidationError from './ValidationError.js';
+
+/**
+ * @typedef {Object} VendingMachine
+ * @property {function} getProducts
+ * @property {function} getChanges
+ * @property {function} getTotalChanges
+ * @property {function} addItem
+ * @property {function} insertCoins
+ */
+
+/**
+ * @typedef {Object} VendingMachineItem
+ * @property {number} index
+ * @property {string} name
+ * @property {number} price
+ * @property {number} amount
+ */
+
+class VendingMachine {
+  #products;
+  #changes;
+
+  reset() {
+    this.#products = [];
+    this.#changes = VENDING_MACHINE_CONSTANT.CHANGES.UNITS.reduce(
+      (result, unit) => ({
+        ...result,
+        [unit]: 0,
+      }),
+      {}
+    );
+  }
+
+  constructor() {
+    this.reset();
+  }
+
+  /**
+   *
+   * @returns {VendingMachineItem[]}
+   */
+  getProducts() {
+    return this.#products;
+  }
+
+  /**
+   *
+   * @returns {Object}
+   */
+  getChanges() {
+    return this.#changes;
+  }
+
+  /**
+   *
+   * @param {VendingMachineItem} vendingMachineItem
+   */
+  addItem({ name, price, amount }) {
+    this.#validateProduct({ name, price, amount });
+    const item = this.#products.find((item) => item.name === name) || {
+      index: this.#products.length,
+      name: removeSpace(name),
+    };
+    const newItem = {
+      ...item,
+      price,
+      amount,
+    };
+    this.#products = [...this.#products.filter((item) => item.name !== name), newItem].sort(
+      (a, b) => a.index - b.index
+    );
+  }
+
+  /**
+   *
+   * @param {VendingMachineItem} vendingMachineItem
+   */
+  #validateProduct({ name, price, amount }) {
+    if (!isNameValid(name)) {
+      throw new ValidationError(ALERT_MESSAGE.VALIDATION.PRODUCT.NAME_BLANK);
+    }
+    if (!isPriceValid(price)) {
+      throw new ValidationError(ALERT_MESSAGE.VALIDATION.PRODUCT.PRICE);
+    }
+    if (!isAmountValid(amount)) {
+      throw new ValidationError(ALERT_MESSAGE.VALIDATION.PRODUCT.AMOUNT);
+    }
```

Comment:
```suggestion
    if (!isNameValid(name)) {
      throw new InvalidProductNameError();
    }
    if (!isPriceValid(price)) {
      throw new InvalidPriceError();
    }
    if (!isAmountValid(amount)) {
      throw new InvalidAmountError();
    }
```
아예 이렇게 별도의 에러로 구분하면 어떨까요?
이럴 경우 테스트하기도 용이하지 않을까요!?

Reply:
용도별로 에러를 구분하고자 했는데
구체화하는 대신에 자판기 미션 다른 부분을 신경써야 할 것 같아서 굳이 오류를 분류하지는 않았습니다.
앱에서 의도한 동작에 관한 오류는 alert 메시지로 출력시키고자 함이었고
나머지 의도치 않은 오류는 alert 메시지로 출력시키지 않고 console.error로 출력시킴으로써
사용성을 생각한 것이었습니다.
이대로 진행해도 될까요?
그렇지 않고 제시해주신 부분으로 별도의 오류로 나누게 된다면 얻을 수 있는 이점으로는 무엇이 있을지 궁금합니다!

Reply:
확장 가능성에 대해 고민해봤을 때,
테스트를 작성할 때 더 직관적으로 판단할 수 있고,
바깥에서 error를 catch할 때도 message로 구분하기보단 class instance로 구분한다면 더 효과적으로 사용할 수 있다고 생각했어요!

Reply:
심도 있게 고민해주신 부분인데 이 부분까지 신경쓰고 가기에는 다른 중요한 부분이 많아서 다 반영하지 못하는 점 양해 부탁드려요.
다만, 다른 미션이나 사이드 프로젝트에서 제안해주신 부분 고민해보면서 코드를 작성할 수 있는 인사이트를 주셔서 소중히 간직하고 있다가 잘 활용해볼게요!